File: exp3/FCFS.java
// this code is of burst time with different arival time 
public class FCFS {

    static void findAvgTime(int bt[], int at[]) {

        int n = bt.length;

        int[] wt = new int[n];
        int[] tat = new int[n];
        int[] ct = new int[n];

        int currentTime = 0;
        int totalWT = 0, totalTAT = 0;

        for (int i = 0; i < n; i++) {

            // If CPU is idle, jump to arrival time
            if (currentTime < at[i]) {
                currentTime = at[i];
            }

            // Waiting time
            wt[i] = currentTime - at[i];

            // Completion time
            ct[i] = currentTime + bt[i];

            // Turnaround time
            tat[i] = ct[i] - at[i];

            // Move current time forward
            currentTime += bt[i];

            totalWT += wt[i];
            totalTAT += tat[i];
        }

        System.out.println("Process\tBT\tAT\tWT\tTAT\tCT");

        for (int i = 0; i < n; i++) {
            System.out.println((i+1) + "\t\t" + bt[i] + "\t" + at[i] +
                               "\t" + wt[i] + "\t" + tat[i] + "\t" + ct[i]);
        }

        System.out.println("\nAverage Waiting Time = " + (double) totalWT / n);
        System.out.println("Average Turnaround Time = " + (double) totalTAT / n);
    }

    public static void main(String[] args) {

        int burst_time[] = {5, 9, 6};
        int arrival_time[] = {0, 3, 6};

        findAvgTime(burst_time, arrival_time);
    }
}

File: exp3/FCFS.cpp
#include <iostream>
using namespace std;

int main() {
    cout << "===== FCFS Scheduling =====\n\n";

    int n = 4;
    int bt[] = {5, 3, 8, 6};
    int wt[4], tat[4];

    wt[0] = 0;

    for(int i = 1; i < n; i++)
        wt[i] = wt[i-1] + bt[i-1];

    for(int i = 0; i < n; i++)
        tat[i] = wt[i] + bt[i];

    cout << "Process\tBT\tWT\tTAT\n";
    for(int i = 0; i < n; i++)
        cout << "P" << i+1 << "\t" << bt[i] << "\t" << wt[i] << "\t" << tat[i] << endl;

    return 0;
}